[FSDP/Checkpoint] Activation offload support in checkpoint_wrapper

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70165

Implements activation offload support in checkpoint_wrapper API via
save_on_cpu hooks. We avoid modifying the torch.utils.checkpoint implementation
and instead compose offload + checkpoint using the save_on_cpu hook for the
former.

Differential Revision: [D33228820](https://our.internmc.facebook.com/intern/diff/D33228820/)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

rohan-varma committed:
[FSDP/Checkpoint] Activation offload support in checkpoint_wrapper

Implements activation offload support in checkpoint_wrapper API via
save_on_cpu hooks. We avoid modifying the torch.utils.checkpoint implementation
and instead compose offload + checkpoint using the save_on_cpu hook for the
former.

Differential Revision: [D33228820](https://our.internmc.facebook.com/intern/diff/D33228820/)

[ghstack-poisoned]

zhaojuanmao commented on torch/distributed/algorithms/_checkpoint/_checkpoint_wrapper.py:
```diff
@@ -17,22 +19,28 @@ def __init__(
-            **kwargs,
-        )
+        offload_mgr = save_on_cpu(pin_memory=True) if self.offload_to_cpu else suppress()
+        with offload_mgr:
```
so basically inner activations are not saved during checkpoint(), save_on_cpu() will just save outer activation tensors on cpu?

rohan-varma replied:
Yeah, inner activations (with the reentrant-based checkpointing impl that's the only one currently supported) will be discarded since we run under torch.no_grad mode. 

All tensors saved via autograd (including those with `save_for_backward` in the checkpointing impl) will be offloaded to CPU as well, and these are the outer activations.

zhaojuanmao commented on test/distributed/fsdp/test_fsdp_checkpoint.py:
```diff
@@ -81,16 +94,18 @@ def _verify_parity(self, losses, outputs, models):
         [CPUOffload(offload_params=True), CPUOffload(offload_params=False)]
     )
-    def test_checkpoint_fsdp_wrapping(self, cpu_offload):
+    @parametrize("offload_activations", [True])
```
nit: [True, False]?

zhaojuanmao reviewed (approved):
lg, thanks!

rohan-varma committed:
Update on "[FSDP/Checkpoint] Activation offload support in checkpoint_wrapper"


Implements activation offload support in checkpoint_wrapper API via
save_on_cpu hooks. We avoid modifying the torch.utils.checkpoint implementation
and instead compose offload + checkpoint using the save_on_cpu hook for the
former.

Differential Revision: [D33228820](https://our.internmc.facebook.com/intern/diff/D33228820/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma commented:
All CI failures are due to docker ` Error: Docker pull failed with exit code 1`.

rohan-varma committed:
Update on "[FSDP/Checkpoint] Activation offload support in checkpoint_wrapper"


Implements activation offload support in checkpoint_wrapper API via
save_on_cpu hooks. We avoid modifying the torch.utils.checkpoint implementation
and instead compose offload + checkpoint using the save_on_cpu hook for the
former.

Differential Revision: [D33228820](https://our.internmc.facebook.com/intern/diff/D33228820/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma committed:
Update on "[FSDP/Checkpoint] Activation offload support in checkpoint_wrapper"


Implements activation offload support in checkpoint_wrapper API via
save_on_cpu hooks. We avoid modifying the torch.utils.checkpoint implementation
and instead compose offload + checkpoint using the save_on_cpu hook for the
former.

Differential Revision: [D33228820](https://our.internmc.facebook.com/intern/diff/D33228820/)

cc pietern mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse @SciPioneer @H-Huang

[ghstack-poisoned]

rohan-varma commented:
CI failure `Build left local git repository checkout dirty` should be unrelated
